DependencyResolution/PostSharp.Samples.DependencyResolution.Contextual/Program.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.ComponentModel.Composition.Hosting;
+using System.ComponentModel.Composition.Primitives;
+using System.Reflection;
+using PostSharp.Aspects;
+using PostSharp.Extensibility;
+using PostSharp.Serialization;
+
+namespace PostSharp.Samples.DependencyResolution.Contextual
+{
+    public interface ILogger
+    {
+        void Log(string message);
+    }
+
+    public static class AspectServiceLocator
+    {
+        private static CompositionContainer container;
+        private static HashSet<object> rules = new HashSet<object>();
+
+        public static void Initialize(ComposablePartCatalog catalog)
+        {
+            container = new CompositionContainer(catalog);
+        }
+
+        public static Lazy<T> GetService<T>(Type aspectType, MemberInfo targetElement) where T : class
+        {
```

Comment:
I'm not sure `Lazy` is the right thing here. We need to cope with the problem of random initialization order, i.e. your aspect may be executed before the catalogue is initialized (especially with logging). One way would be to return an object of type say `Service<T>` where `service.Value` would return `null` before initialization, but after initialization, the value would be cached.